### PR TITLE
Refactored TaacAlgorithm and added TaacLAlgorithm and TaacQAlgorithm

### PR DESCRIPTION
**Pull request by hnyu**
1. Refactored and renamed TASAC to TAAC
2. Implemented piecewise quadratic/linear trajectory policy algorithms (TaacQ and TaacL)
3. Added a "benchmarks" directory under `alf/examples` for common task configurations.

After refactoring, TAAC didn't have any performance difference (tested on BipedalWalker and Fetch), compared to the arxiv paper.

### CODE REVIEW

File: alf/algorithms/taac_algorithm.py [hunk 1]
```diff
@@ -505,80 +497,101 @@ def _calc_critic_loss(self, info: TasacInfo):
...
         if mode == Mode.predict:
-            action = dist_utils.epsilon_greedy_sample(action_dist,
-                                                      epsilon_greedy)
+            b1_a = dist_utils.epsilon_greedy_sample(b1_a_dist, epsilon_greedy)
         else:
-            action = dist_utils.rsample_action_distribution(action_dist)
+            b1_a = dist_utils.rsample_action_distribution(b1_a_dist)
+
+        b0_tau = self._update_tau(state.tau)
+        # This should be a determinisitc function converting b1_a to b1_tau
```
Comment (Haichao-Zhang) [inline alf/algorithms/taac_algorithm.py]: determinisitc -> deterministic

**Comment by hnyu** [comment]
I'm currently implementing the piecewise linear trajectory algorithm. Will update this PR once finished.

**Comment by hnyu** [comment]
Updated with PltpAlgorithm and performance.

**Comment by hnyu** [comment]
Made a comparison between squashing and clipping the updated `a` and `v`, and found that clipping is both simpler and better. So replaced the original version. 

taac_fetch

Also rename PLTP -> TaacL and PQTP -> TaacQ

File: alf/algorithms/taac_algorithm.py [hunk 2]
```diff
@@ -686,7 +726,173 @@ def calc_loss(self, info: TasacInfo):
...
+        args = args + (tau_mask, )
+        return self._make_networks_impl(*args)
+
+    def _update_tau(self, tau):
+        """Return a constant trajectory."""
+        return tau
+
+    def _action2tau(self, a, tau):
+        """Return a constant trajectory."""
+        return Tau(a=a, v=a, u=a)
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: why not `Tau(a=a)`?

File: alf/algorithms/taac_algorithm.py [hunk 3]
```diff
@@ -686,7 +726,173 @@ def calc_loss(self, info: TasacInfo):
...
+        """
+        a = torch.clamp(tau.a + 2. * tau.v, min=-1., max=1.)
+        return tau._replace(a=a)
+
+    def _action2tau(self, a, tau):
+        """Given a new action at the next step and the current traj ``tau``, infer
+        the new traj's first derivative.
+        """
+        v = (a - tau.a) / 2.
+        return Tau(a=a, v=v, u=v)
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: Why not `Tau(a=a, v=v)`?

File: alf/algorithms/taac_algorithm.py [hunk 4]
```diff
@@ -686,7 +726,173 @@ def calc_loss(self, info: TasacInfo):
...
+
+    def _action2tau(self, a, tau):
+        """Given a new action at the next step and the current traj ``tau``, infer
+        the new traj's second derivative, assuming resetting ``tau.v`` to 0 first.
+        """
+        v = (a - tau.a) / 2.
+        u = v / 2.
+        return Tau(a=a, v=v, u=u)
+
+    def _tau2action(self, tau):
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: Seems that `_tau2action()` for Taac, TaacL, TaacQ are same. Why do you bother to implement them in  the subclasses?

File: alf/algorithms/taac_algorithm.py [hunk 5]
```diff
@@ -323,14 +330,15 @@ def __init__(self,
         log_alpha = (nn.Parameter(torch.zeros(())),
                      nn.Parameter(torch.zeros(())))
 
-        train_state_spec = TasacState(
+        train_state_spec = TaacState(
+            tau=Tau(a=action_spec, v=action_spec, u=action_spec),
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: this can be set according to tau_mask so that v and u can be () if the mask is False.

File: alf/algorithms/taac_algorithm.py [hunk 6]
```diff
@@ -386,22 +393,42 @@ def __init__(self,
...
-                self._num_critic_replicas * reward_spec.numel)
-
-        input_preprocessors = (None,
-                               EmbeddingPreprocessor(
-                                   input_tensor_spec=action_spec,
-                                   embedding_dim=observation_spec.numel))
-        # computes the action probability conditioned on s and a^-
+            return net.make_parallel(self._num_critic_replicas)
+
+        tau_spec = nest.map_structure(lambda _: action_spec, Tau())
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: if tau_mask is False, use () instead of action_spec

File: alf/algorithms/taac_algorithm.py [hunk 7]
```diff
@@ -386,22 +393,42 @@ def __init__(self,
...
+                alf.layers.FC(action_spec.numel, observation_spec.numel)),
+            Tau())
+
         actor_network = actor_network_cls(
-            input_tensor_spec=(observation_spec, action_spec),
-            input_preprocessors=input_preprocessors,
-            preprocessing_combiner=nest_utils.NestConcat(),
+            input_tensor_spec=(observation_spec, tau_spec),
+            input_preprocessors=(alf.layers.Detach(), tau_embedding),
+            preprocessing_combiner=nest_utils.NestConcat(
```
Comment (emailweixu) [inline alf/algorithms/taac_algorithm.py]: if tau_mask is used to construct tau_spec, it will be needed here.

File: alf/algorithms/taac_algorithm.py [hunk 8]
```diff
@@ -18,6 +18,7 @@
 import torch
 import torch.nn as nn
 import torch.distributions as td
+import numpy as np
```
Comment (Haichao-Zhang) [inline alf/algorithms/taac_algorithm.py]: can move this line before Line18, for alphabetic ordering

File: alf/examples/benchmarks/fetch/fetch_conf.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+"""
+
+from functools import partial
+
+import alf
+from alf.environments import suite_robotics
+from alf.utils.math_ops import clipped_exp
+from alf.algorithms.data_transformer import RewardNormalizer
+from alf.optimizers import AdamTF
+
```
Comment (Haichao-Zhang) [inline alf/examples/benchmarks/fetch/fetch_conf.py]: can order these imports

**Review by Haichao-Zhang — APPROVED** [review]

**Review by emailweixu — APPROVED** [review]
